crates/inverted/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+#![allow(clippy::len_without_is_empty)]
+
+use base::distance::DistanceKind;
+use base::index::{IndexOptions, SearchOptions};
+use base::operator::Borrowed;
+use base::operator::Operator;
+use base::scalar::{ScalarLike, F32};
+use base::search::{Collection, Element, Payload, Source, Vectors};
+use base::vector::{VectorBorrowed, VectorKind};
+use common::dir_ops::sync_dir;
+use common::json::Json;
+use common::mmap_array::MmapArray;
+use common::remap::RemappedCollection;
+use quantization::{operator::OperatorQuantization, Quantization};
+use storage::{OperatorStorage, Storage};
+
+use std::collections::BTreeMap;
+use std::fs::create_dir;
+use std::path::Path;
+
+pub trait OperatorInverted: Operator + OperatorQuantization + OperatorStorage {}
+
+impl<T: Operator + OperatorQuantization + OperatorStorage> OperatorInverted for T {}
+
+#[allow(dead_code)]
+pub struct Inverted<O: OperatorInverted> {
+    storage: O::Storage,
+    quantization: Quantization<O>,
+    payloads: MmapArray<Payload>,
+    indexes: Json<Vec<u32>>,
+    offsets: Json<Vec<u32>>,
+    scores: Json<Vec<F32>>,
+}
+
+impl<O: OperatorInverted> Inverted<O> {
+    pub fn create(path: impl AsRef<Path>, options: IndexOptions, source: &impl Source<O>) -> Self {
+        if options.vector.v != VectorKind::SVecf32 {
+            panic!("inverted index only supports `SVecf32` vectors");
+        }
+        if options.vector.d != DistanceKind::Dot {
+            panic!("inverted index only supports `Dot` distance");
+        }
```

Comment:
It should be checked in `validate`. Putting checks here will make building index fail silently.